File: src/ui.js
"use strict";
const readline = require("readline/promises");
const { stdin, stdout } = require("process");
const chalk = require("chalk");
const { marked } = require("marked");
const { AGENTS_DIRS } = require("./config.js");
const fs = require("fs");
const path = require("path");

const SLASH_COMMANDS = ["/plan", "/build", "/mode", "/skill", "/skills", "/unskill", "/agent", "/agents", "/default", "/clear", "/help", "/exit"];

const PROMPT_EMOJIS = { plan: "?", build: ">" };

function buildCompletions() {
  const words = [...SLASH_COMMANDS];
  const seen = new Set(SLASH_COMMANDS);
  for (const d of AGENTS_DIRS) {
    if (fs.existsSync(d)) {
      for (const f of fs.readdirSync(d).filter((x) => x.endsWith(".md")).sort()) {
        const name = "@" + f.slice(0, -3);
        if (!seen.has(name)) {
          seen.add(name);
          words.push(name);
        }
      }
    }
  }
  return words;
}

function completer(line, allWords) {
  const trimmed = line.trimStart();
  if (!trimmed.startsWith("/") && !trimmed.startsWith("@")) return [[], line];
  const matches = allWords.filter((w) => w.startsWith(trimmed));
  return [matches, line];
}

let _promptMode = "plan";

function setPromptMode(mode) {
  _promptMode = mode;
}

function reloadCompletions() {
  return buildCompletions();
}

function panel(title, content, borderColor = "blue") {
  const borderChars = { blue: chalk.blue, green: chalk.green, red: chalk.red, yellow: chalk.yellow, dim: chalk.dim };
  const b = borderChars[borderColor] || chalk.blue;
  const titleStr = title ? ` ${title} ` : "";
  const contentLines = content.split("\n");
  const maxInner = Math.max(
    titleStr.length,
    ...contentLines.map((l) => stripAnsi(l).length)
  );
  const width = Math.min(maxInner + 4, 80);
  const borderLine = b("\u2500".repeat(width));
  let result = `${borderLine}\n`;
  if (title) result += `${b(title)}\n`;
  for (const l of contentLines) {
    result += `${b("\u2502")} ${l.padEnd(width - 2)}\n`;
  }
  result += borderLine;
  return result;
}

function stripAnsi(str) {
  return str.replace(/\x1B\[[0-9;]*[mK]/g, "");
}

function showHeader(mode, model) {
  console.log(panel("DWCode", `Model: ${chalk.bold(model)}          Mode: ${chalk.green("● " + mode.toUpperCase())}`, "blue"));
}

function showStatus(msg) {
  console.log(panel("...", msg, "dim"));
}

function showInfo(msg) {
  console.log(panel("i", msg, "dim"));
}

function showError(msg) {
  console.log(chalk.red(panel("Error", String(msg), "red")));
}

function renderMarkdown(text) {
  try {
    if (require("marked-terminal")) {
      return marked(text, { renderer: require("marked-terminal") });
    }
  } catch {}
  return text;
}

async function getInput(allWords) {
  const rl = readline.createInterface({
    input: stdin,
    output: stdout,
    completer: (line) => completer(line, allWords),
  });

  const emoji = PROMPT_EMOJIS[_promptMode] || ">";
  try {
    const answer = await rl.question(`${emoji} ${_promptMode} > `);
    return answer.trim();
  } catch {
    return null;
  } finally {
    rl.close();
  }
}

module.exports = {
  setPromptMode,
  reloadCompletions,
  showHeader,
  showStatus,
  showInfo,
  showError,
  getInput,
  renderMarkdown,
  panel,
  SLASH_COMMANDS,
};

File: ui.py
from rich.console import Console
from rich.markdown import Markdown
from rich.panel import Panel
from rich.text import Text
from prompt_toolkit import PromptSession
from prompt_toolkit.history import InMemoryHistory
from prompt_toolkit.styles import Style as PTStyle
from prompt_toolkit.completion import Completer, Completion
from prompt_toolkit.key_binding import KeyBindings
from prompt_toolkit.filters import has_completions
from config import AGENTS_DIRS

console = Console()

SLASH_COMMANDS = ["/plan", "/build", "/mode", "/skill", "/skills", "/unskill", "/agent", "/agents", "/default", "/clear", "/help", "/exit"]

def _build_completions():
    words = list(SLASH_COMMANDS)
    seen = set()
    for d in AGENTS_DIRS:
        if d.exists():
            for f in sorted(d.glob("*.md")):
                name = "@" + f.stem
                if name not in seen:
                    seen.add(name)
                    words.append(name)
    return words

ALL_WORDS = _build_completions()

def reload_completions():
    ALL_WORDS[:] = _build_completions()

class TriggerCompleter(Completer):
    def get_completions(self, document, complete_event):
        word = document.get_word_before_cursor(WORD=True)
        if not word or not word.startswith(('/', '@')):
            return
        for w in ALL_WORDS:
            if w.startswith(word):
                yield Completion(w, start_position=-len(word))

pt_style = PTStyle.from_dict({
    "prompt": "ansicyan bold",
})

kb = KeyBindings()

@kb.add("enter", filter=has_completions)
def _(event):
    b = event.app.current_buffer
    if b.complete_state:
        b.complete_state = None

@kb.add("enter", filter=~has_completions)
def _(event):
    event.app.current_buffer.validate_and_handle()

_session = None
_prompt_mode = "plan"

def _get_session():
    global _session
    if _session is None:
        _session = PromptSession(
            history=InMemoryHistory(),
            style=pt_style,
            completer=TriggerCompleter(),
            complete_while_typing=True,
            key_bindings=kb,
        )
    return _session

def set_prompt_mode(mode):
    global _prompt_mode
    _prompt_mode = mode

PROMPT_EMOJIS = {"plan": "🤔", "build": "🔧"}

def get_input():
    try:
        emoji = PROMPT_EMOJIS.get(_prompt_mode, "❯")
        return _get_session().prompt(f"{emoji} {_prompt_mode} > ").strip()
    except (EOFError, KeyboardInterrupt):
        return None

def show_header(mode, model):
    console.print(Panel(
        f"Model: [bold]{model}[/bold]          Mode: [green]● {mode.upper()}[/green]",
        title="[bold]DWCode[/bold]",
        border_style="blue",
        padding=(0, 1),
    ))

class AssistantStream:
    def __init__(self):
        self.text = ""

    def __enter__(self):
        return self

    def __exit__(self, *args):
        if self.text:
            console.print(Panel(
                Markdown(self.text.strip()),
                title="[bold blue]DWCode[/bold blue]",
                border_style="blue",
            ))

    def update_text(self, delta):
        self.text += delta

def show_status(msg):
    console.print(Panel(
        Text(msg),
        title="⏳",
        border_style="dim",
        padding=(0, 1),
    ))

def show_info(msg):
    console.print(Panel(
        Text(msg),
        title="ℹ️",
        border_style="dim",
        padding=(0, 1),
    ))

def show_error(msg):
    console.print(Panel(
        Text(str(msg), style="red bold"),
        title="[bold red]Error[/bold red]",
        border_style="red",
        padding=(0, 1),
    ))